src/lib/libdb1/man/hash.3
.\"	$ssdlinux: hash.3,v 1.1 2004/11/11 17:13:56 kanoh Exp $
.\"	$NetBSD: hash.3,v 1.6 1996/05/03 21:26:50 cgd Exp $
.\"
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)hash.3	8.6 (Berkeley) 8/18/94
.\"
.TH HASH 3 "August 18, 1994"
.UC 7
.SH NAME
hash \- hash database access method
.SH SYNOPSIS
.nf
.ft B
#include <sys/types.h>
#include <db.h>
.ft R
.fi
.SH DESCRIPTION
The routine
.IR dbopen
is the library interface to database files.
One of the supported file formats is hash files.
The general description of the database access methods is in
.IR dbopen (3),
this manual page describes only the hash specific information.
.PP
The hash data structure is an extensible, dynamic hashing scheme.
.PP
The access method specific data structure provided to
.I dbopen
is defined in the <db.h> include file as follows:
.sp
typedef struct {
.RS
u_int bsize;
.br
u_int ffactor;
.br
u_int nelem;
.br
u_int cachesize;
.br
u_int32_t (*hash)(const void *, size_t);
.br
int lorder;
.RE
} HASHINFO;
.PP
The elements of this structure are as follows:
.TP
bsize
.I Bsize
defines the hash table bucket size, and is, by default, 256 bytes.
It may be preferable to increase the page size for disk-resident tables
and tables with large data items.
.TP
ffactor
.I Ffactor
indicates a desired density within the hash table.
It is an approximation of the number of keys allowed to accumulate in any
one bucket, determining when the hash table grows or shrinks.
The default value is 8.
.TP
nelem
.I Nelem
is an estimate of the final size of the hash table.
If not set or set too low, hash tables will expand gracefully as keys
are entered, although a slight performance degradation may be noticed.
The default value is 1.
.TP
cachesize
A suggested maximum size, in bytes, of the memory cache.
This value is
.B only
advisory, and the access method will allocate more memory rather
than fail.
.TP
hash
.I Hash
is a user defined hash function.
Since no hash function performs equally well on all possible data, the
user may find that the built-in hash function does poorly on a particular
data set.
User specified hash functions must take two arguments (a pointer to a byte
string and a length) and return a 32-bit quantity to be used as the hash
value.
.TP
lorder
The byte order for integers in the stored database metadata.
The number should represent the order as an integer; for example, 
big endian order would be the number 4,321.
If
.I lorder
is 0 (no order is specified) the current host order is used.
If the  file already exists, the specified value is ignored and the
value specified when the tree was created is used.
.PP
If the file already exists (and the O_TRUNC flag is not specified), the
values specified for the parameters bsize, ffactor, lorder and nelem are
ignored and the values specified when the tree was created are used.
.PP
If a hash function is specified,
.I hash_open
will attempt to determine if the hash function specified is the same as
the one with which the database was created, and will fail if it is not.
.PP
Backward compatible interfaces to the routines described in
.IR dbm (3),
and
.IR ndbm (3)
are provided, however these interfaces are not compatible with
previous file formats.
.SH ERRORS
The
.I hash
access method routines may fail and set
.I errno
for any of the errors specified for the library routine
.IR dbopen (3).
.SH "SEE ALSO"
.IR btree (3),
.IR dbopen (3),
.IR mpool (3),
.IR recno (3)
.sp
.IR "Dynamic Hash Tables" ,
Per-Ake Larson, Communications of the ACM, April 1988.
.sp
.IR "A New Hash Package for UNIX" ,
Margo Seltzer, USENIX Proceedings, Winter 1991.
.SH BUGS
Only big and little endian byte order is supported.


src/ChangeLog
$ssdlinux: ChangeLog,v 1.492 2005/03/20 09:18:33 kanoh Exp $

Sun Mar 20 18:15:57 JST 2005  Tamotsu Kanoh <kanoh@kanoh.org>
	* Update GCC to version 3.3.5.

Fri Mar 18 11:27:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.20.

Fri Mar 18 09:57:30 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050318 Release.

Wed Mar 16 17:00:49 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update strace to version 4.5.10.

Wed Mar 16 15:48:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update postfix to version 2.2.1.

Tue Mar 15 11:32:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update gettext to version 0.14.3.

Tue Mar 15 10:18:52 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update psmisc to version 21.6.

Mon Mar 14 12:48:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Updated bind to version 9.2.5.

Mon Mar 14 10:15:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050312.

Fri Mar 11 16:43:48 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update tnftp to version 20050103.

Fri Mar 11 09:58:54 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050311 Release.

Thu Mar 10 17:44:49 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update postfix to version 2.2.0.

Thu Mar 10 16:11:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update iptables to version 1.3.1.

Thu Mar 10 14:14:08 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update openssh to version 4.0p1.

Mon Mar  7 10:15:38 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050305.

Fri Mar  4 18:12:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050304 Release.

Wed Mar  2 15:04:15 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update e2fsprogs to version 1.36.

Wed Mar  2 10:45:01 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update dhcp to version 3.0.2.

Mon Feb 28 10:14:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050226.

Fri Feb 25 16:23:38 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ex to version 050225.

Fri Feb 25 16:05:29 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update gettext to version 0.14.2.

Fri Feb 25 09:46:36 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050225 Release.

Mon Feb 21 11:58:44 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update zlib to version 1.2.2.

Mon Feb 21 11:28:09 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.18.

Mon Feb 21 10:57:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050219.

Fri Feb 18 10:39:24 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050218 Release.

Tue Feb 15 18:33:14 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050215 Release.

Mon Feb 14 12:03:37 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update iptables to version 1.3.0.

Mon Feb 14 11:03:45 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update file to version 4.13.

Mon Feb 14 10:40:03 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update libtool to version 1.5.14.

Mon Feb 14 10:33:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update automake to version 1.9.5.

Mon Feb 14 10:28:42 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050212.

Thu Feb 10 18:02:01 JST 2005  Jun Todoroki <todoroki@plathome.co.jp>
	* Update usbustils to version 0.70 & add libusb-0.1.9.

Wed Feb  9 11:34:42 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update grub to version 0.96.

Wed Feb  9 14:36:01 JST 2005  Jun Todoroki <todoroki@plathome.co.jp>
	* Update madwifi to version 20050209.

Mon Feb  7 14:20:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update strace to version 4.5.9.

Mon Feb  7 13:47:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sudo to version 1.6.8p7.

Mon Feb  7 11:17:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update cvs to version 1.11.19.

Mon Feb  7 10:47:05 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update libtool to version 1.5.12.

Mon Feb  7 10:38:15 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050205.

Mon Jan 31 12:29:41 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sed to version 4.1.4.

Mon Jan 31 12:19:56 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050129.

Mon Jan 31 11:29:54 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.15.

Sat Jan 29 18:05:07 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050129 Release.

Thu Jan 27 15:57:02 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update util-linux to version 2.12q.

Thu Jan 27 12:41:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update shadow to version 4.0.7.

Thu Jan 27 11:38:48 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update procps to version 3.2.5.
	
Thu Jan 27 11:18:56 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ex to version 050119.

Thu Jan 27 11:01:33 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update dump to version 0.4b39.

Thu Jan 27 10:36:16 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sed to version 4.1.3.

Wed Jan 26 17:08:50 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050122.

Fri Jan 21 23:18:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050121 Release.

Wed Jan 19 11:18:33 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update a patch for sendmail (Errata: 2005-01-17).

Mon Jan 17 11:04:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
        * Apply a patch for sendmail (Errata: 2005-01-15).


src/usr.bin/bmake/hash.c
/*	$ssdlinux: hash.c,v 1.1.1.1 2002/05/02 13:37:10 kanoh Exp $	*/
/*	$NetBSD: hash.c,v 1.9 2000/06/11 07:54:32 mycroft Exp $	*/

/*
 * Copyright (c) 1988, 1989, 1990 The Regents of the University of California.
 * Copyright (c) 1988, 1989 by Adam de Boor
 * Copyright (c) 1989 by Berkeley Softworks
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Adam de Boor.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef MAKE_BOOTSTRAP
static char rcsid[] = "$NetBSD: hash.c,v 1.9 2000/06/11 07:54:32 mycroft Exp $";
#else
#include <sys/cdefs.h>
#ifndef lint
#if 0
static char sccsid[] = "@(#)hash.c	8.1 (Berkeley) 6/6/93";
/*
#else
__RCSID("$NetBSD: hash.c,v 1.9 2000/06/11 07:54:32 mycroft Exp $");
*/
#endif
#endif /* not lint */
#endif

/* hash.c --
 *
 * 	This module contains routines to manipulate a hash table.
 * 	See hash.h for a definition of the structure of the hash
 * 	table.  Hash tables grow automatically as the amount of
 * 	information increases.
 */
#include "sprite.h"
#include "make.h"
#include "hash.h"

/*
 * Forward references to local procedures that are used before they're
 * defined:
 */

static void RebuildTable __P((Hash_Table *));

/*
 * The following defines the ratio of # entries to # buckets
 * at which we rebuild the table to make it larger.
 */

#define rebuildLimit 3

/*
 *---------------------------------------------------------
 *
 * Hash_InitTable --
 *
 *	This routine just sets up the hash table.
 *
 * Results:
 *	None.
 *
 * Side Effects:
 *	Memory is allocated for the initial bucket area.
 *
 *---------------------------------------------------------
 */

void
Hash_InitTable(t, numBuckets)
	register Hash_Table *t;	/* Structure to use to hold table. */
	int numBuckets;		/* How many buckets to create for starters.
				 * This number is rounded up to a power of
				 * two.   If <= 0, a reasonable default is
				 * chosen. The table will grow in size later
				 * as needed. */
{
	register int i;
	register struct Hash_Entry **hp;

	/*
	 * Round up the size to a power of two.
	 */
	if (numBuckets <= 0)
		i = 16;
	else {
		for (i = 2; i < numBuckets; i <<= 1)
			 continue;
	}
	t->numEntries = 0;
	t->size = i;
	t->mask = i - 1;
	t->bucketPtr = hp = (struct Hash_Entry **)emalloc(sizeof(*hp) * i);
	while (--i >= 0)
		*hp++ = NULL;
}

/*
 *---------------------------------------------------------
 *
 * Hash_DeleteTable --
 *
 *	This routine removes everything from a hash table
 *	and frees up the memory space it occupied (except for
 *	the space in the Hash_Table structure).
 *
 * Results:
 *	None.
 *
 * Side Effects:
 *	Lots of memory is freed up.
 *
 *---------------------------------------------------------
 */

void
Hash_DeleteTable(t)
	Hash_Table *t;
{
	register struct Hash_Entry **hp, *h, *nexth = NULL;
	register int i;

	for (hp = t->bucketPtr, i = t->size; --i >= 0;) {
		for (h = *hp++; h != NULL; h = nexth) {
			nexth = h->next;
			free((char *)h);
		}
	}
	free((char *)t->bucketPtr);

	/*
	 * Set up the hash table to cause memory faults on any future access
	 * attempts until re-initialization.
	 */
	t->bucketPtr = NULL;
}

/*
 *---------------------------------------------------------
 *
 * Hash_FindEntry --
 *
 * 	Searches a hash table for an entry corresponding to key.
 *
 * Results:
 *	The return value is a pointer to the entry for key,
 *	if key was present in the table.  If key was not
 *	present, NULL is returned.
 *
 * Side Effects:
 *	None.
 *
 *---------------------------------------------------------
 */

Hash_Entry *
Hash_FindEntry(t, key)
	Hash_Table *t;		/* Hash table to search. */
	char *key;		/* A hash key. */
{
	register Hash_Entry *e;
	register unsigned h;
	register char *p;

	for (h = 0, p = key; *p;)
		h = (h << 5) - h + *p++;
	p = key;
	for (e = t->bucketPtr[h & t->mask]; e != NULL; e = e->next)
		if (e->namehash == h && strcmp(e->name, p) == 0)
			return (e);
	return (NULL);
}

/*
 *---------------------------------------------------------
 *
 * Hash_CreateEntry --
 *
 *	Searches a hash table for an entry corresponding to
 *	key.  If no entry is found, then one is created.
 *
 * Results:
 *	The return value is a pointer to the entry.  If *newPtr
 *	isn't NULL, then *newPtr is filled in with TRUE if a
 *	new entry was created, and FALSE if an entry already existed
 *	with the given key.
 *
 * Side Effects:
 *	Memory may be allocated, and the hash buckets may be modified.
 *---------------------------------------------------------
 */

Hash_Entry *
Hash_CreateEntry(t, key, newPtr)
	register Hash_Table *t;	/* Hash table to search. */
	char *key;		/* A hash key. */
	Boolean *newPtr;	/* Filled in with TRUE if new entry created,
				 * FALSE otherwise. */
{
	register Hash_Entry *e;
	register unsigned h;
	register char *p;
	int keylen;
	struct Hash_Entry **hp;

	/*
	 * Hash the key.  As a side effect, save the length (strlen) of the
	 * key in case we need to create the entry.
	 */
	for (h = 0, p = key; *p;)
		h = (h << 5) - h + *p++;
	keylen = p - key;
	p = key;
	for (e = t->bucketPtr[h & t->mask]; e != NULL; e = e->next) {
		if (e->namehash == h && strcmp(e->name, p) == 0) {
			if (newPtr != NULL)
				*newPtr = FALSE;
			return (e);
		}
	}

	/*
	 * The desired entry isn't there.  Before allocating a new entry,
	 * expand the table if necessary (and this changes the resulting
	 * bucket chain).
	 */
	if (t->numEntries >= rebuildLimit * t->size)
		RebuildTable(t);
	e = (Hash_Entry *) emalloc(sizeof(*e) + keylen);
	hp = &t->bucketPtr[h & t->mask];
	e->next = *hp;
	*hp = e;
	e->clientData = NULL;
	e->namehash = h;
	(void) strcpy(e->name, p);
	t->numEntries++;

	if (newPtr != NULL)
		*newPtr = TRUE;
	return (e);
}

/*
 *---------------------------------------------------------
 *
 * Hash_DeleteEntry --
 *
 * 	Delete the given hash table entry and free memory associated with
 *	it.
 *
 * Results:
 *	None.
 *
 * Side Effects:
 *	Hash chain that entry lives in is modified and memory is freed.
 *
 *---------------------------------------------------------
 */

void
Hash_DeleteEntry(t, e)
	Hash_Table *t;
	Hash_Entry *e;
{
	register Hash_Entry **hp, *p;

	if (e == NULL)
		return;
	for (hp = &t->bucketPtr[e->namehash & t->mask];
	     (p = *hp) != NULL; hp = &p->next) {
		if (p == e) {
			*hp = p->next;
			free((char *)p);
			t->numEntries--;
			return;
		}
	}
	(void) write(2, "bad call to Hash_DeleteEntry\n", 29);
	abort();
}

/*
 *---------------------------------------------------------
 *
 * Hash_EnumFirst --
 *	This procedure sets things up for a complete search
 *	of all entries recorded in the hash table.
 *
 * Results:
 *	The return value is the address of the first entry in
 *	the hash table, or NULL if the table is empty.
 *
 * Side Effects:
 *	The information in searchPtr is initialized so that successive
 *	calls to Hash_Next will return successive HashEntry's
 *	from the table.
 *
 *---------------------------------------------------------
 */

Hash_Entry *
Hash_EnumFirst(t, searchPtr)
	Hash_Table *t;			/* Table to be searched. */
	register Hash_Search *searchPtr;/* Area in which to keep state
					 * about search.*/
{
	searchPtr->tablePtr = t;
	searchPtr->nextIndex = 0;
	searchPtr->hashEntryPtr = NULL;
	return Hash_EnumNext(searchPtr);
}

/*
 *---------------------------------------------------------
 *
 * Hash_EnumNext --
 *    This procedure returns successive entries in the hash table.
 *
 * Results:
 *    The return value is a pointer to the next HashEntry
 *    in the table, or NULL when the end of the table is
 *    reached.
 *
 * Side Effects:
 *    The information in searchPtr is modified to advance to the
 *    next entry.
 *
 *---------------------------------------------------------
 */

Hash_Entry *
Hash_EnumNext(searchPtr)
	register Hash_Search *searchPtr; /* Area used to keep state about
					    search. */
{
	register Hash_Entry *e;
	Hash_Table *t = searchPtr->tablePtr;

	/*
	 * The hashEntryPtr field points to the most recently returned
	 * entry, or is nil if we are starting up.  If not nil, we have
	 * to start at the next one in the chain.
	 */
	e = searchPtr->hashEntryPtr;
	if (e != NULL)
		e = e->next;
	/*
	 * If the chain ran out, or if we are starting up, we need to
	 * find the next nonempty chain.
	 */
	while (e == NULL) {
		if (searchPtr->nextIndex >= t->size)
			return (NULL);
		e = t->bucketPtr[searchPtr->nextIndex++];
	}
	searchPtr->hashEntryPtr = e;
	return (e);
}

/*
 *---------------------------------------------------------
 *
 * RebuildTable --
 *	This local routine makes a new hash table that
 *	is larger than the old one.
 *
 * Results:
 * 	None.
 *
 * Side Effects:
 *	The entire hash table is moved, so any bucket numbers
 *	from the old table are invalid.
 *
 *---------------------------------------------------------
 */

static void
RebuildTable(t)
	register Hash_Table *t;
{
	register Hash_Entry *e, *next = NULL, **hp, **xp;
	register int i, mask;
        register Hash_Entry **oldhp;
	int oldsize;

	oldhp = t->bucketPtr;
	oldsize = i = t->size;
	i <<= 1;
	t->size = i;
	t->mask = mask = i - 1;
	t->bucketPtr = hp = (struct Hash_Entry **) emalloc(sizeof(*hp) * i);
	while (--i >= 0)
		*hp++ = NULL;
	for (hp = oldhp, i = oldsize; --i >= 0;) {
		for (e = *hp++; e != NULL; e = next) {
			next = e->next;
			xp = &t->bucketPtr[e->namehash & mask];
			e->next = *xp;
			*xp = e;
		}
	}
	free((char *)oldhp);
}
